## Part 1: Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df

Age Attrition     BusinessTravel              Department  \
0      41       Yes      Travel_Rarely                   Sales   
1      49        No  Travel_Frequently  Research & Development   
2      37       Yes      Travel_Rarely  Research & Development   
3      33        No  Travel_Frequently  Research & Development   
4      27        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
1465   36        No  Travel_Frequently  Research & Development   
1466   39        No      Travel_Rarely  Research & Development   
1467   27        No      Travel_Rarely  Research & Development   
1468   49        No  Travel_Frequently                   Sales   
1469   34        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                    1          2  Life Sciences                        2   
1                    8          1  Life Sciences                        3   
2                    2          2          Other                        4   
3                    3          4  Life Sciences                        4   
4                    2          1        Medical                        1   
...                ...        ...            ...                      ...   
1465                23          2        Medical                        3   
1466                 6          1        Medical                        4   
1467                 4          3  Life Sciences                        2   
1468                 2          3        Medical                        4   
1469                 8          3        Medical                        2   

      HourlyRate  JobInvolvement  ...  PerformanceRating  \
0             94               3  ...                  3   
1             61               2  ...                  4   
2             92               2  ...                  3   
3             56               3  ...                  3   
4             40               3  ...                  3   
...          ...             ...  ...                ...   
1465          41               4  ...                  3   
1466          42               2  ...                  3   
1467          87               4  ...                  4   
1468          63               2  ...                  3   
1469          82               4  ...                  3   

     RelationshipSatisfaction  StockOptionLevel TotalWorkingYears  \
0                           1                 0                 8   
1                           4                 1                10   
2                           2                 0                 7   
3                           3                 0                 8   
4                           4                 1                 6   
...                       ...               ...               ...   
1465                        3                 1                17   
1466                        1                 1                 9   
1467                        2                 1                 6   
1468                        4                 0                17   
1469                        1                 0                 6   

      TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                         0               1               6   
1                         3               3              10   
2                         3               3               0   
3                         3               3               8   
4                         3               3               2   
...                     ...             ...             ...   
1465                      3               3               5   
1466                      5               3               7   
1467                      0               3               6   
1468                      3               2               9   
1469                      3               4  

In [2]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [3]:
attrition_df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EnvironmentSatisfaction', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'NumCompaniesWorked', 'OverTime', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [4]:
# Create y_df with the Attrition and Department columns

y_df = attrition_df[['Attrition', 'Department']]

In [5]:
y_df['Attrition'].value_counts()

Attrition
No     1233
Yes     237
Name: count, dtype: int64

In [6]:
y_df['Department'].value_counts()

Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64

In [7]:
# The Attrition and Department parameters are not balanced.

In [8]:
# Create a list of at least 10 column names to use as X data

X_selected = ['Age', 'DistanceFromHome', 'YearsAtCompany', 'YearsSinceLastPromotion', 'Education', 'JobSatisfaction', 'OverTime', 'StockOptionLevel', 'WorkLifeBalance',
        'NumCompaniesWorked', 'MaritalStatus']

# Create X_df using your selected columns
X_df = attrition_df[X_selected]

# Show the data types for X_df
X_df.dtypes


Age                         int64
DistanceFromHome            int64
YearsAtCompany              int64
YearsSinceLastPromotion     int64
Education                   int64
JobSatisfaction             int64
OverTime                   object
StockOptionLevel            int64
WorkLifeBalance             int64
NumCompaniesWorked          int64
MaritalStatus              object
dtype: object

In [9]:
# Split the data into training and testing sets
#X_train, X_test, y_train, y_test = train_test_split(X_df, y_df)
# 

In [10]:
X_df['OverTime'].value_counts()

OverTime
No     1054
Yes     416
Name: count, dtype: int64

In [11]:
#Collect non-numerics

non_num_df = X_df.select_dtypes(include=object)

for column in non_num_df.columns:
    print(non_num_df[column].value_counts())

OverTime
No     1054
Yes     416
Name: count, dtype: int64
MaritalStatus
Married     673
Single      470
Divorced    327
Name: count, dtype: int64


In [12]:
non_num_df.columns

Index(['OverTime', 'MaritalStatus'], dtype='object')

In [13]:
# This datafram requires any order needed. OneHotEncoder would be suitable.


In [14]:
# Encode OverTime
le_O = LabelEncoder()
X_df['OverTime'] = le_O.fit_transform(X_df['OverTime'])



/var/folders/s7/d5dl_ywn63zg7wx1h9wckr2r0000gn/T/ipykernel_5058/2097289980.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_df['OverTime'] = le_O.fit_transform(X_df['OverTime'])


In [15]:
X_df

Age  DistanceFromHome  YearsAtCompany  YearsSinceLastPromotion  \
0      41                 1               6                        0   
1      49                 8              10                        1   
2      37                 2               0                        0   
3      33                 3               8                        3   
4      27                 2               2                        2   
...   ...               ...             ...                      ...   
1465   36                23               5                        0   
1466   39                 6               7                        1   
1467   27                 4               6                        0   
1468   49                 2               9                        0   
1469   34                 8               4                        1   

      Education  JobSatisfaction  OverTime  StockOptionLevel  WorkLifeBalance  \
0             2                4         1                 0                1   
1             1                2         0                 1                3   
2             2                3         1                 0                3   
3             4                3         1                 0                3   
4             1                2         0                 1                3   
...         ...              ...       ...               ...              ...   
1465          2                4         0                 1                3   
1466          1                1         0                 1                3   
1467          3                2         1                 1                3   
1468          3                2         0                 0                2   
1469          3                3         0                 0                4   

      NumCompaniesWorked MaritalStatus  
0                      8        Single  
1                      1       Married  
2                      6        Single  
3                      1       Married  
4                      9       Married  
...                  ...           ...  
1465                   4       Married  
1466                   4       Married  
1467                   1       Married  
1468                   2       Married  
1469                   2       Married  

[1470 rows x 11 columns]

In [16]:
# Encode Married

le_M = LabelEncoder()
X_df['MaritalStatus'] = le_M.fit_transform(X_df['MaritalStatus'])

/var/folders/s7/d5dl_ywn63zg7wx1h9wckr2r0000gn/T/ipykernel_5058/4107626349.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_df['MaritalStatus'] = le_M.fit_transform(X_df['MaritalStatus'])


In [17]:
X_df

Age  DistanceFromHome  YearsAtCompany  YearsSinceLastPromotion  \
0      41                 1               6                        0   
1      49                 8              10                        1   
2      37                 2               0                        0   
3      33                 3               8                        3   
4      27                 2               2                        2   
...   ...               ...             ...                      ...   
1465   36                23               5                        0   
1466   39                 6               7                        1   
1467   27                 4               6                        0   
1468   49                 2               9                        0   
1469   34                 8               4                        1   

      Education  JobSatisfaction  OverTime  StockOptionLevel  WorkLifeBalance  \
0             2                4         1                 0                1   
1             1                2         0                 1                3   
2             2                3         1                 0                3   
3             4                3         1                 0                3   
4             1                2         0                 1                3   
...         ...              ...       ...               ...              ...   
1465          2                4         0                 1                3   
1466          1                1         0                 1                3   
1467          3                2         1                 1                3   
1468          3                2         0                 0                2   
1469          3                3         0                 0                4   

      NumCompaniesWorked  MaritalStatus  
0                      8              2  
1                      1              1  
2                      6              2  
3                      1              1  
4                      9              1  
...                  ...            ...  
1465                   4              1  
1466                   4              1  
1467                   1              1  
1468                   2              1  
1469                   2              1  

[1470 rows x 11 columns]

In [18]:
y_df['Department'].value_counts()

Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64

In [19]:
y_df['Attrition'].value_counts()

Attrition
No     1233
Yes     237
Name: count, dtype: int64

In [20]:
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df)

In [21]:
department_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore', drop=None)

department_encoder.fit(y_train[['Department']])

y_department_train = department_encoder.transform(y_train[['Department']])
y_department_test = department_encoder.transform(y_test[['Department']])

y_department_train, y_department_test



(array([[0., 0., 1.],
        [0., 1., 0.],
        [0., 0., 1.],
        ...,
        [0., 1., 0.],
        [0., 1., 0.],
        [0., 0., 1.]]),
 array([[0., 1., 0.],
        [0., 1., 0.],
        [0., 0., 1.],
        ...,
        [0., 1., 0.],
        [0., 1., 0.],
        [0., 0., 1.]]))

In [22]:
attrition_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore', drop=None)

attrition_encoder.fit(y_train[['Attrition']])

y_attr_train = attrition_encoder.transform(y_train[['Attrition']])
y_attr_test = attrition_encoder.transform(y_test[['Attrition']])

y_attr_train, y_attr_test

(array([[1., 0.],
        [1., 0.],
        [1., 0.],
        ...,
        [1., 0.],
        [1., 0.],
        [1., 0.]]),
 array([[1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [0., 1.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [0., 1.],
        [1., 0.],
        [1., 0.],
        [1., 0.],
        [1., 

In [23]:
# encoder_dept = OneHotEncoder(handle_unknown='ignore', sparse_output=False)
# encoded_dept = encoder_dept.fit_transform(y_df[['Department']])
# department_columns = encoder_dept.get_feature_names_out(['Department'])
# df_encoded_dept = pd.DataFrame(encoded_dept, columns=department_columns)

# encoder_attr = OneHotEncoder(handle_unknown='ignore', sparse_output=False)
# encoded_attr = encoder_attr.fit_transform(y_df[['Attrition']])
# attrition_columns = encoder_attr.get_feature_names_out(['Attrition'])
# df_encoded_attr = pd.DataFrame(encoded_attr, columns=attrition_columns)

# y_processed_df = pd.concat([ 
#     df_encoded_dept,
#     df_encoded_attr
#     ], axis=1)

# y_processed_df

In [24]:
#X_train, X_test, y_train, y_test = train_test_split(X_df, y_processed_df)

In [25]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
scaler.fit(X_train)
# Scale the training and testing data

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Create, Compile, and Train the Model

In [26]:
# Create a OneHotEncoder for the Department column
input_nodes = len(X_df.columns)

# Fit the encoder to the training data
input_layer = layers.Input(shape=(input_nodes,), name='input_layer')

# Create two new variables by applying the encoder
shared1 = layers.Dense(64, activation='relu')(input_layer)
shared2 = layers.Dense(128, activation='relu')(shared1)
# to the training and testing data

In [27]:
Department_hidden = layers.Dense(32, activation='relu')(shared2)

Department_output = layers.Dense(3, activation='sigmoid', name='Department_out')(Department_hidden)

In [28]:
# Create attrition branch
# Create the hidden layer
Attrition_hidden = layers.Dense(32, activation='relu')(shared2)

# Create the output layer
Attrition_output = layers.Dense(2, activation='sigmoid', name='Attrition_out')(Attrition_hidden)

In [29]:
# Create the model
model = Model(inputs=input_layer, outputs=[Department_output, Attrition_output])

# Compile
model.compile(optimizer='adam',
                loss={'Department_out': 'categorical_crossentropy', 'Attrition_out': 'binary_crossentropy'},
                metrics={'Department_out': 'accuracy', 'Attrition_out': 'accuracy'})

model.summary()

Model: "functional_1"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 11)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense (Dense)       │ (None, 64)        │        768 │ input_layer[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_1 (Dense)     │ (None, 128)       │      8,320 │ dense[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_2 (Dense)     │ (None, 32)        │      4,128 │ dense_1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_3 (Dense)     │ (None, 32)        │      4,128 │ dense_1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Department_out      │ (None, 3)         │         99 │ dense_2[0][0]     │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Attrition_out       │ (None, 2)         │         66 │ dense_3[0][0]     │
│ (Dense)             │                   │            │                   │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 17,509 (68.39 KB)

 Trainable params: 17,509 (68.39 KB)

 Non-trainable params: 0 (0.00 B)

In [30]:
# Train the model
model.fit(X_train_scaled,
            {'Department_out': y_department_train, 'Attrition_out': y_attr_train},
            epochs=100,
            batch_size=32,
            validation_split=0.2)

Epoch 1/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - Attrition_out_accuracy: 0.7859 - Department_out_accuracy: 0.5031 - loss: 1.5306 - val_Attrition_out_accuracy: 0.8326 - val_Department_out_accuracy: 0.6561 - val_loss: 1.3301
Epoch 2/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_out_accuracy: 0.8324 - Department_out_accuracy: 0.6483 - loss: 1.2237 - val_Attrition_out_accuracy: 0.8326 - val_Department_out_accuracy: 0.6561 - val_loss: 1.2807
Epoch 3/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - Attrition_out_accuracy: 0.8220 - Department_out_accuracy: 0.6531 - loss: 1.1483 - val_Attrition_out_accuracy: 0.8416 - val_Department_out_accuracy: 0.6561 - val_loss: 1.2311
Epoch 4/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_out_accuracy: 0.8357 - Department_out_accuracy: 0.6398 - loss: 1.1267 - val_Attrition_out_accuracy: 0.8552 - val_Department_out_accuracy: 0.6380 - val_loss: 1.2172
Epoch 5/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_out_accuracy: 0.8532 - Depart

In [31]:
model.metrics_names

['loss', 'compile_metrics']

In [32]:
# Evaluate the model
test_results = model.evaluate(X_test_scaled, {'Department_out': y_department_test, 'Attrition_out': y_attr_test})
print(test_results)

12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 549us/step - Attrition_out_accuracy: 0.8336 - Department_out_accuracy: 0.5501 - loss: 4.4451
[4.2285332679748535, 0.8288043737411499, 0.5869565010070801]


In [33]:
# Print the accuracy for both department and attrition
print(f"Attrition predictions accuracy: {test_results[1]:.3f}")
print(f"Department predictions accuracy: {test_results[2]:.3f}")

Attrition predictions accuracy: 0.829
Department predictions accuracy: 0.587


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?
     
2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

# Summary
YOUR ANSWERS HERE

1.  Not very accurate.  For instance, Department predictions accuracy is only around 58% versus Attrition accuracy  
    83%. 
2.  The activation function I chose was relu for hidden layers and sigmoid for the output. 
3.  I could have tried to picked a different X parameters to try to see if that improves.